lightning/src/offers/signer.rs

```diff
@@ -74,6 +80,7 @@ impl Metadata {
 			// derived, as wouldn't be the case if a Metadata::Bytes with length PaymentId::LENGTH +
 			// Nonce::LENGTH had been set explicitly.
 			Metadata::Bytes(bytes) => bytes.len() == PaymentId::LENGTH + Nonce::LENGTH,
+			Metadata::Nonce(_) => false,
```

Comment:
Possibly, yes. It's a little more complicated as the context would also need the `Nonce` and we'd need another variant here that wraps a 48-byte array (so it could be used with `as_ref`) containing the *encrypted* payment id and nonce. It would avoid increasing the size of a `Refund` with a path. Though for an `InvoiceRequest` the size savings isn't necessary.

Reply:
Mostly for the reasons given in the rest of the sentence. :) Though, originally I thought the whole change would be more complicated until I switched to the additional enum variant approach.

Another consideration is we actually *do* need to set the payer metadata to something for both the `InvoiceRequest` and `Refund` cases according to the spec.

```
The writer:
  - MUST set `invreq_metadata` to an unpredictable series of bytes.

The reader:
  - MUST fail the request if `invreq_payer_id` or `invreq_metadata` are not present.
```

For the case where we include blinded paths, we could set *only* the 16-byte `Nonce` instead? Or maybe the first 16 bytes of the encrypted `PaymentId` to avoid allocating another `Vec`? We wouldn't need it for anything ourself, as both the `PaymentId` and `Nonce` would come back in the `BlindedPath`. But the spec wants it because:

```
  - if `offer_node_id` is present, and `invreq_metadata` is identical to a previous `invoice_request`:
    - MAY simply reply with the previous invoice.
```

And:

```
Note: the `invreq_metadata` is numbered 0 (not in the
80-159 range for other invreq fields) as this is the first
TLV element, which ensures payer-provided entropy is used in hashing
for [Signature Calculation](#signature-calculation).
```

Reply:
Ok, I made this change except I've left the `payer_metadata` as is for now. When trying to change it to only include the nonce, `ln::offers_tests::pays_bolt12_invoice_asynchronously` fails. Presumably because the code is trying to re-verify the invoice using `payer_metadata`. I'll need to put some thought in if we could do this.

Reply:
Not in this situation because the `payment_id` is encrypted in the `payer_metadata`. So an attacker wouldn't be able to include the unencrypted `payment_id` in any blinded path that they composed unless they learned of it by some other means.

I think the solution for asynchronous handling is to either include the `OfferContext` in `Event::InvoiceReceived` or drop the verification check from `ChannelManager::send_payment_for_bolt12_invoice`. Then we could remove the encrypted `payment_id` from the `payer_metadata` to save those 32 bytes, which is more important for a `Refund`. We'd still need to include some random data as per the spec, so we'd leave the nonce.

Reply:
> Not in this situation because the `payment_id` is encrypted in the `payer_metadata`. So an attacker wouldn't be able to include the unencrypted `payment_id` in any blinded path that they composed unless they learned of it by some other means.

Discussed offline. Ended up dropping the nonce from the payer metadata when including blinded paths.
 
> I think the solution for asynchronous handling is to either include the `OfferContext` in `Event::InvoiceReceived` or drop the verification check from `ChannelManager::send_payment_for_bolt12_invoice`. Then we could remove the encrypted `payment_id` from the `payer_metadata` to save those 32 bytes, which is more important for a `Refund`. We'd still need to include some random data as per the spec, so we'd leave the nonce.

Included the `OffersContext` in `Event::InvoiceReceived`.